go/vt/vtgate/plan_execute.go

```diff
@@ -158,10 +162,19 @@ func (e *Executor) newExecute(
 			return err
 		}
 
+		// 6: Retry if needed.
 		rootCause := vterrors.RootCause(err)
 		if rootCause != nil && strings.Contains(rootCause.Error(), "enforce denied tables") {
 			log.V(2).Infof("Retry: %d, will retry query %s due to %v", try, query, err)
-			lastVSchemaCreated = vs.GetCreated()
+			if try == 0 { // We are going to retry at least once
```

Comment:
Can you explain why this deferred clearing of plans when the `executor.newExecutor()` returns helps? 

When we retry, we wait for a newer vschema. Shouldn't we clear the plans before executing the query with the new vschema.

Reply:
> Can you explain why this deferred clearing of plans when the executor.newExecutor() returns helps?

As noted in the description, this is the part the prevented the query shape from failing indefinitely and causing a never-ending cycle of keyspace events when there was a ~ concurrent buffer event.

The executor's VschemaManager clears the plans when it receives a new vschema via the watcher (it calls executor.SaveVSchema() in its watches subscriber callback). This happens concurrently with the KeyspaceEventWatcher also receiving the new vschema and in its subscriber callback processing it (which includes getting info on all shards from the topo), and eventually determining that the keyspace is consistent and ending the buffering window. So there was a race with query retries such that a query could be planned against the wrong side just as the keyspace event is getting resolved and the buffers drained. Then that bad plan is the cached plan for the query until you do another `RebuildSrvVSchema`.

> When we retry, we wait for a newer vschema. Shouldn't we clear the plans before executing the query with the new vschema.

As I noted above, the VSchemaManager already clears the plans when a new vschema is received. We wait for a newer vschema, but not indefinitely as the wait times out and we retry without it. So if we timeout just as the new vschema is coming in, we could plan with the old vschema, putting that into the just cleared plan cache for all subsequent executions. Do you remember when you, @harshit-gangal, and I were discussing this on Zoom? That's where this idea originated and you and I were both confused by how the query failures continued even after the buffering window ended due to hitting the max duration.

Reply:
I improved the behavior and comments here: https://github.com/vitessio/vitess/pull/15701/commits/7675cf88b97d9404a7e775887f041be6faff36d9

Sorry for not explaining this more clearly ahead of time! 🙏

Reply:
This is good.